File: java/try-amazon-kinesis/src/main/java/com/github/mindpin/try_amazon_kinesis/util/ShardUtils.java
package com.github.mindpin.try_amazon_kinesis.util;

import com.amazonaws.auth.AWSCredentials;
import com.amazonaws.services.kinesis.AmazonKinesis;
import com.amazonaws.services.kinesis.AmazonKinesisClient;
import com.amazonaws.services.kinesis.model.DescribeStreamRequest;
import com.amazonaws.services.kinesis.model.DescribeStreamResult;
import com.amazonaws.services.kinesis.model.Shard;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by li fei on 2015/10/8.
 */
public class ShardUtils {
    public static List<Shard> get_shards(AmazonKinesis client, String steam_name){
        DescribeStreamRequest describeStreamRequest = new DescribeStreamRequest();
        describeStreamRequest.setStreamName(steam_name);
        List<Shard> shards = new ArrayList<Shard>();
        String exclusiveStartShardId = null;
        do {
            describeStreamRequest.setExclusiveStartShardId( exclusiveStartShardId );
            DescribeStreamResult describeStreamResult = client.describeStream( describeStreamRequest );
            shards.addAll( describeStreamResult.getStreamDescription().getShards() );
            if (describeStreamResult.getStreamDescription().getHasMoreShards() && shards.size() > 0) {
                exclusiveStartShardId = shards.get(shards.size() - 1).getShardId();
            } else {
                exclusiveStartShardId = null;
            }
        } while ( exclusiveStartShardId != null );

        System.out.println("shares size: " + shards.size());
        return shards;
    }

    public static void shard_info(Shard shard){
        System.out.println("shard id : " + shard.getShardId());
        System.out.println("shard startingHashKey : " + shard.getHashKeyRange().getStartingHashKey());
        System.out.println("shard endingHashKey : " + shard.getHashKeyRange().getEndingHashKey());
    }
}

File: ruby/try-amazon-kinesis/lib/shard_utils.rb
class ShardUtils
  class << self
    def get_shards(client, stream_name)
      shards = []
      has_more_shard = true
      exclusive_start_shard_id = nil

      while has_more_shard do
        resp = client.describe_stream({
          stream_name: stream_name,
          exclusive_start_shard_id: exclusive_start_shard_id
        })
        _shards = resp.stream_description.shards
        shards += _shards

        has_more_shard = resp.stream_description.has_more_shards
        exclusive_start_shard_id = _shards.last.shard_id
      end

      p "shares size: #{shards.count}"
      shards
    end

    def shard_info(shard)
      p "shard_id: #{shard.shard_id}"
      p "starting_hash_key: #{shard.hash_key_range.starting_hash_key}"
      p "ending_hash_key: #{shard.hash_key_range.ending_hash_key}"
    end
  end
end
